cmds/field/field.go

```diff
@@ -0,0 +1,366 @@
+	}
+	return r
+}
+
+func parsefield(lex *lexer) (int, toktype) {
+	typ := lex.peektype()
+	if typ == TokNF {
+		lex.next()
+		return nf, TokNF
+	}
+	return parsenum(lex), TokNum
+}
+
+func parsenum(lex *lexer) int {
+	tok, typ := lex.next()
+	if typ == TokEOF {
+		fatal("EOF in number parser")
+	}
+	if typ == TokNum {
+		num, _ := strconv.Atoi(tok)
+		return num
+	}
+	if typ != TokDash {
+		fatal("number parser error: unexpected token '%v'", tok)
+	}
+	tok, typ = lex.next()
+	if typ == TokEOF {
+		fatal("negative number parse error: unexpected EOF")
+	}
+	if typ != TokNum {
+		fatal("number parser error: bad lexical token '%v'", tok)
+	}
+	num, _ := strconv.Atoi(tok)
+	return -num
+}
+
+func isdelim(typ toktype) bool {
+	if typ == TokComma || typ == TokPipe || typ == TokSpace {
+		return true
+	}
+	return false
+}
+
+type toktype int
+
+const (
+	TokError = iota
+	TokEOF
+	TokComma
+	TokPipe
+	TokDash
+	TokNum
+	TokSpace
+	TokNF
+
+	eof = -1
+)
+
+type lexer struct {
+	input string
+	tok   string
+	typ   toktype
+	start int
+	pos   int
+	width int
+}
+
+func (lex *lexer) peek() (string, toktype) {
+	return lex.tok, lex.typ
+}
+
+func (lex *lexer) peektype() toktype {
+	return lex.typ
+}
+
+func (lex *lexer) peektok() string {
+	return lex.tok
+}
+
+func (lex *lexer) next() (string, toktype) {
+	tok, typ := lex.peek()
+	lex.tok, lex.typ = lex.scan()
+	return tok, typ
+}
+
+func (lex *lexer) scan() (string, toktype) {
+	switch r := lex.nextrune(); {
+	case r == eof:
+		return "", TokEOF
+	case r == ',':
+		return lex.token(), TokComma
+	case r == '|':
+		return lex.token(), TokPipe
+	case r == '-':
+		return lex.token(), TokDash
+	case r == 'N':
+		lex.consume()
+		r = lex.nextrune()
+		if r == 'F' {
+			lex.consume()
+		}
+		lex.ignore()
+		return lex.token(), TokNF
+	case unicode.IsDigit(r):
+		for r := lex.nextrune(); unicode.IsDigit(r); r = lex.nextrune() {
+			lex.consume()
+		}
+		lex.ignore()
+		return lex.token(), TokNum
+	case unicode.IsSpace(r):
+		for r := lex.nextrune(); unicode.IsSpace(r); r = lex.nextrune() {
+			lex.consume()
+		}
+		lex.ignore()
+		return lex.token(), TokSpace
+	default:
+		fatal("Lexical error at character '%v'", r)
+	}
+	return "", TokError
+}
+
+func (lex *lexer) nextrune() (r rune) {
+	if lex.pos >= len(lex.input) {
+		lex.width = 0
+		return eof
+	}
+	r, lex.width = utf8.DecodeRuneInString(lex.input[lex.pos:])
+	return r
+}
+
+func (lex *lexer) consume() {
+	lex.pos += lex.width
+	lex.width = 0
+}
+
+func (lex *lexer) ignore() {
+	lex.width = 0
+}
+
+func (lex *lexer) token() string {
+	lex.consume()
+	tok := lex.input[lex.start:lex.pos]
+	lex.start = lex.pos
+	return tok
+}
+
+func process(file *os.File, rv []frange, delim *regexp.Regexp, outsep string, collapse bool) {
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		prefix := ""
+		printed := false
+		line := scanner.Text()
+		fields := split(line, delim, collapse)
+		for _, r := range rv {
+			begin, end := r.begin, r.end
+			if begin == 0 {
```

Comment:
could these next three things be a switch on begin? also, the `begin--` can be a bit confusing, I feel like it'd be easier to just set the value you want in each case (and not set `begin = 1` in the 0 case just so it goes back to 0...)

```go
switch {
case begin == 0:
  pprefix(prefix)
  prefix = outsep
  fmt.Print(line)
  printed = true
case begin == nf:
  begin = len(fields) - 1
case begin < 0:
  begin += len(fields)
default:
  begin--
}
```